cmd/kube-scheduler/app/options/options.go

```diff
@@ -222,6 +222,15 @@ func (o *Options) ApplyTo(c *schedulerappconfig.Config) error {
 		c.ComponentConfig = *cfg
 	}
 
+	// Build kubeconfig first to so that if it fails, it doesn't cause leaking
+	// goroutines (started from initializing secure serving - which underneath
```

Comment:
LGTM, but can't we add a context to that queue that we can cancel?

Reply:
what queue?

Reply:
> what queue?

It's coming from SecureServingOptions::ApplyTo:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/serving.go#L266
which internally creates: DynamicServingContent:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_serving_content.go#L67
[which creates a queue]

But the trick is that any queue created by any component would behave exactly the same.


> but can't we add a context to that queue that we can cancel?

In theory we can add that, but usage of Queue is widely spread across the whole codebase:
https://github.com/kubernetes/kubernetes/blob/ee0b179a7b9867203a9cc73f5c90909318a240e3/staging/src/k8s.io/client-go/util/workqueue/queue.go
and that would be a huge refactoring.
The additional problem is that most of our codebase is structured in a way:
- a method creating a certain object/controller
- `Run()` method that gets the context/stopCh that then closes everything it should 

The problem is when the creation method is spawning some goroutines, which is exactly the case for Queue, because Queue is initialized on creation (FWIW - it's also the case for almost every single controller we have - we just don't exercise the case when controllers are created by `Run()` method is not called, except from one testcase in `test/integration/serving`.

I fully agree it's not beautiful, but changing all of that would be on its own huge effort and I'm not convinced it's worth it.

Reply:
The Shutdown() method is called from Run() methods in all controllers.
The problem are those patological scenarios tested via this one test that is failing (and testing whether it fails) before Run() method is called.

We could in theory add a bunch of shutdowns everywhere - but that would be extremely ugly and error prone imho.

Reply:
> I agree that we need a shorter-term solution (this PR).
> 
> I notice that the workqueue has a Shutdown method. We can possibly call that somewhere without having to introduce a context. It would have a lot of repeated code though :(

/hold cancel

the refactor is not likely going to happen soon 😄 
let´s move with the short term solution